crates/biome_js_type_info/src/flattening/intersections.rs

```diff
@@ -0,0 +1,319 @@
+    Namespace,
+    Never,
+    Object,
+    Primitive,
+    Unknown,
+}
+
+impl MergedTypeKind {
+    fn intersection_with(self, other: Self) -> Self {
+        match (self, other) {
+            (Self::Any, _) | (_, Self::Any) => Self::Any,
+            (Self::Class, Self::Class) => Self::Class,
+            (Self::Class, Self::Function | Self::Interface | Self::Object | Self::Namespace)
+            | (Self::Function | Self::Interface | Self::Object | Self::Namespace, Self::Class) => {
+                Self::Class
+            }
+            (Self::Function, Self::Function) => Self::Function,
+            (Self::Interface, Self::Interface) => Self::Interface,
+            (Self::Interface, Self::Function) | (Self::Function, Self::Interface) => {
+                Self::Interface
+            }
+            (Self::Namespace, Self::Namespace) => Self::Namespace,
+            (Self::Namespace, Self::Function | Self::Object | Self::Interface)
+            | (Self::Function | Self::Object | Self::Interface, Self::Namespace) => Self::Namespace,
+            (Self::Never, some) | (some, Self::Never) => some,
+            (Self::Object, Self::Object) => Self::Object,
+            (Self::Object, Self::Function | Self::Interface)
+            | (Self::Function | Self::Interface, Self::Object) => Self::Object,
+            (Self::Primitive, Self::Primitive) => Self::Never,
+            (Self::Primitive, _) | (_, Self::Primitive) => Self::Primitive,
+            (Self::Unknown, _) | (_, Self::Unknown) => Self::Unknown,
+        }
+    }
+}
+
+impl MergedType {
+    fn from_type(ty: TypeData, resolver: &dyn TypeResolver) -> Self {
+        fn from_type_with_static(
+            ty: TypeData,
+            is_static: bool,
+            resolver: &dyn TypeResolver,
+        ) -> MergedType {
+            let filter_static = |member: &TypeMember| -> bool {
+                if is_static {
+                    member.is_static()
+                } else {
+                    !member.is_static()
+                }
+            };
+
+            // FIXME: We're throwing away info such as prototypes for now...
+            match ty {
+                TypeData::AnyKeyword | TypeData::Conditional => MergedType::Any,
+                TypeData::BigInt
+                | TypeData::Boolean
+                | TypeData::Null
+                | TypeData::Number
+                | TypeData::String
+                | TypeData::Symbol
+                | TypeData::Undefined => MergedType::Primitive(ty),
+                TypeData::Class(class) => {
+                    MergedType::Class(class.members.into_iter().filter(filter_static).collect())
+                }
+                TypeData::Function(function) => MergedType::Function(function),
+                TypeData::InstanceOf(instance) => match resolver.resolve_and_get(&instance.ty) {
+                    Some(resolved_data) => {
+                        from_type_with_static(resolved_data.to_data(), false, resolver)
```

Comment:
This is recursion, and I suppose it's intentional. Should we have a threshold to avoid going too deep?

Reply:
Good catch! I'll just refuse to go instances to other instances, since those should be flattened anyway.